### ISSUE TITLE
Ticketline codes

### ISSUE THREAD

**Issue by terrlundamo** [Skatteetaten/saf-t#4]
Hi,

In the element <ctLine> there is an element <lineType> that refers to ticketline codes:
"Reference to ticketline codes. Description of the
code must be stated in the table 'Basics'"

Where can i find these codes? I've found all other codes here on github, but not these. 
(Predefined basic ID = 5 for ticketline codes)

**Comment by steinarberggren** [comment]
There are no predefined codes for BasicType 05 - Ticketline codes. Only listed examples in the technical documentation. The system specific codes, are to be listed in basics. The predefinedBasicID will have no values. 

The documentation is to be updated with information on that lineType (mandatory) as a minimum must be filled with codes (system specific) that makes it possible to separate lines representing sale from lines representing corrections and others. We might add a codelist (predfinedBasicID) for this purpose in future revisions of the format. 

Please se the description on page 7 and the example showing the same situation for BasicType 08.
http://www.skatteetaten.no/globalassets/standardformat-regnskapsaf-t/norwegian-saf-t-cash-register-data---documentation.pdf

**Comment by terrlundamo** [comment]
Ok thank you.

**Comment by terrlundamo** [comment]
I see correction is listed as an example lineType. What is a correction in a ctLine? 
Is it the same as corrections in the "reportCorrLines" element?
For intance number of a product decreased from 5 to 4, product removed before receiving payment etc?

**Comment by steinarberggren** [comment]
Yes, the reportCorrLines is the summary on the Z/X-report of the corrections made on tcLine, shown by the lineType.  In addition to the example you mentioned above, it can also be a "correct last line" (korriger siste) line. 

The sum of the ctLines representing sales line types, adjusted for lines representing corrections that reduces  the sales, must match the sum of the cashTransaction. It is important that all the ctLines are included, regardless of the lineType.

**Comment by terrlundamo** [comment]
Is it the sum of the sales lines that must match the sum of the cashTransaction? And the the correction lines are just extra information that is not calculated against the cashTransaction sum?

If the correction lines is included in the calculation i don't understand how the sums will match.
Example:
A product with a unit price of 50 Kr have decreased it's amount twice in the same order, once from 5 to 4, and once from 10 to 1, the ctLine with the sales type will have a sum equal to one of that product (50). The two correction lines will have a combined negative sum equal to 10 products (500). Adding those up gives -450, while the cash transaction sum is 50.

**Comment by steinarberggren** [comment]
As I understand your example, the lines are changed. Not keept the same, and written new lines. 

Does this example make sense? 
		
			
![line corrections](https://cloud.githubusercontent.com/assets/16304420/20488299/cfdba6dc-b006-11e6-9233-df4a9d76a13a.png)

**Comment by terrlundamo** [comment]
I dont know if i misunderstand, or we are talking about different things.

My example is a single purchase where the only line on the receipt is a single product with a cost of 50kr. Before the customer paid, the number of items was altered four times:

- Operator adds a single product (artID 123) to "shopping cart"  (Not logged in journal)
- Amount changed by operator from 1 to 5 (Not logged, since only decrease is logged)
- Amount changed by operator from 5 to 4 (Logged in journal)
- Amount changed by operator from 4 to 10 (Not logged, since only decrease is logged)
- Amount changed by operator from 10 to 1 (Logged in journal)
- Customer pay 50 Kr, transaction is finalized and receipt printed.

In your example the first line is SALE with qnt 5, where does this information come from if it is never logged? There is never a sale of 5 articles.

From the documentation: " If, for instance, the transaction describes a sales slip, the ctLines are the slip lines".

Is the corrections in your example changes done to the transaction after it was paid?

**Comment by steinarberggren** [comment]
The example reflext changes done before the transaction is executed/paid. What has to be clearified is the requirements for logging the changes. Please see the regulations: 
- Section 1-2. Definitions, k) continuous use. 
- Section 2-7. Electronic journal. 
- Section In Section 2-8-2. X report s) the number of line corrections specified by type and amount

As this question is regarding the legal side, not a technical issue alone, we need to consult the legal departement before we post an answer. Please post a feedback if you would like us to clearify this, and we will consult the legal departement.

**Comment by terrlundamo** [comment]
Ok. I must have misunderstood the definitions regarding products added to shopping cart. 
In section 1-2 k) it says that sales should be logged, i assumed this meant the transaction, not adding products to shopping cart.

I can't find any description on the logic behind the cashtransaction element regarding corrections and sales. I have the technical description of course, but this only describes the format of the XML file.
What will my example look like if a discount is given when there's 10 articles in the shopping cart, before qty is reduced from 10 to 1?
Is that a correction? Does it change the sum of the sale line? 

Secondly, the documentation says: "If, for instance, the transaction describes a sales slip, the ctLines are the slip lines". This means that the receipt the customer receives will have 4 lines, 2 sale and 2 corrections?

Thanks

**Comment by terrlundamo** [comment]
Clarification:
The sale lines adjusted for the correction lines must match the sum of the cashtransaction, how will discount affect the summation? The cashtransaction sum will be different, so either sale lines or correction lines must change, i assume it's not a correction line, but how will this example look like?

Maybe it will become very clear once the example files are made available.

**Comment by terrlundamo** [comment]
The example file did not cover this, so i have to ask again. How would a cashtransaction element look like when a discount is given to the article, example:

1. adds article to shopping cart (50 kr)
2. increases qty from 1 to 10
3. gives 50% discount
4. decreases qty from 10 to 1

Something like this?:
- Sale 50 kr
- Sale 450 Kr
- What happens when 50% discount is added? Sum of shopping cart is now 225 Kr
- Correction 200 Kr

**Comment by terrlundamo** [comment]
Hi again,

Ref our meeting yesterday, I've been thinking about what we discussed of adding ctLine type "discount".
Since discount is already part of each individual ctLine, it doesn't seem to be correct to add a separate ctLine for discount.

Originally increases in qty was not supposed to be logged, so how about this solution:
- ctLines are the products that are paid for in the transaction, not increases/decreases of qty.
- Discounts are included in each ctLine
- Any corrections that are made will be logged as an event, not as a ctLine

It's quite clearly stated in the the saf-t documentation: "If, for instance, the transaction describes a sales slip, the ctLines are the slip lines"
